histolab/tiler.py

```diff
@@ -1121,3 +1121,288 @@ def _scores(
+    suffix : str, optional
+        Suffix to be added to the tile filename. Default is '.png'
+    mpp : float, optional.
+        Micron per pixel resolution. If provided, takes precedence over level.
+        Default is None.
+    """
+
+    def __init__(
+        self,
+        scorer: Scorer,
+        tile_size: Tuple[int, int],
+        threshold: float = 0.0,
+        level: int = 0,
+        check_tissue: bool = True,
+        tissue_percent: float = 80.0,
+        pixel_overlap: int = 0,
+        prefix: str = "",
+        suffix: str = ".png",
+        mpp: float = None,
+    ):
+        self.scorer = scorer
+        self.threshold = threshold
+
+        super().__init__(
+            tile_size,
+            level,
+            check_tissue,
+            tissue_percent,
+            pixel_overlap,
+            prefix,
+            suffix,
+            mpp=mpp,
+        )
+
+    def extract(
+        self,
+        slide: Slide,
+        extraction_mask: BinaryMask = BiggestTissueBoxMask(),
+        report_path: str = None,
+        log_level: str = "INFO",
+    ) -> None:
+        """Extract grid tiles and save them to disk, according to a scoring function and
+        following this filename pattern:
+        `{prefix}tile_{tiles_counter}_level{level}_{x_ul_wsi}-{y_ul_wsi}-{x_br_wsi}-{y_br_wsi}{suffix}`
+
+        Save a CSV report file with the saved tiles and the associated score.
+
+        Parameters
+        ----------
+        slide : Slide
+            Slide from which to extract the tiles
+        extraction_mask : BinaryMask, optional
+            BinaryMask object defining how to compute a binary mask from a Slide.
+            Default `BiggestTissueBoxMask`.
+        report_path : str, optional
+            Path to the CSV report. If None, no report will be saved
+        log_level: str, {"DEBUG", "INFO", "WARNING", "ERROR", "CRITICAL"}
+            Threshold level for the log messages. Default "INFO"
+
+        Raises
+        ------
+        TileSizeError
+            If the tile size is larger than the slide size
+        LevelError
+            If the level is not available for the slide
+        """
+        level = logging.getLevelName(log_level)
+        logger.setLevel(level)
+        self._validate_level(slide)
+        self.tile_size = self._tile_size(slide)
+        self.pixel_overlap = int(self._scale_factor(slide) * self.pixel_overlap)
+        self._validate_tile_size(slide)
+
+        filtered_tiles_scores, filtered_tiles_scaled_scores = self._tiles_generator(
+            slide, extraction_mask
+        )
+
+        tiles_counter = 0
+        filenames = []
+
+        for tiles_counter, (score, tile_wsi_coords) in enumerate(filtered_tiles_scores):
+            tile = slide.extract_tile(
+                tile_wsi_coords,
+                tile_size=self.final_tile_size,
+                mpp=self.mpp,
+                level=self.level if self.mpp is None else None,
+            )
```

Comment:
IMHO the tile should be extracted in the `_tiles_generator` and then iterating over the tiles as we do in the `GridTiler`.